Adding getting started instructions for GKE, Istio, and Kgateway

nicolexin opened the pull request:
Update inference extension getting started guide:
- Add instructions for GKE, Istio and Kgateway
- Remove instructions for Envoy Gateway

liu-cong commented:
/ok-to-test

robscott commented:
Thanks @nicolexin!

For KGateway:
/cc @danehans @christian-posta 

For Istio:
/cc @LiorLieberman

rramkumar1 commented on config/manifests/gateway/istio/resources.yaml:
```diff
@@ -0,0 +1,37 @@
+    matches:
+    - path:
+        type: PathPrefix
+        value: /completion
```
I think this needs to be /v1/completion right?

Or just make the path "/"

LiorLieberman replied:
IIRC should be `/v1/completions`

rramkumar1 replied:
Ah yep, probably just making this "/" is easier.

nicolexin replied:
Make sense! I have updated to use "/".

LiorLieberman commented on site-src/guides/index.md:
```diff
@@ -56,55 +55,126 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+
+      1. Install Istio
+      
+      Follow the Istio installation guide https://istio.io/latest/docs/setup/install/
```
note - this will need to be slightly different as the release to be from a branch, and not master. will update once it is released

LiorLieberman commented on config/manifests/gateway/istio/resources.yaml:
```diff
@@ -0,0 +1,37 @@
+  name: llm-route
+spec:
+  hostnames:
+  - foo.example.com
```
this is leftovers from experiments, can be removed.

LiorLieberman commented on config/manifests/gateway/istio/resources.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: llm-route
```
if we keep httproutes separate, should we align them with the same name? Alternatively, we can collapse them to one route shared between implementations (as discussed offline)

LiorLieberman commented on config/manifests/gateway/kgateway/resources.yaml:
```diff
@@ -0,0 +1,41 @@
+    - path:
+        type: PathPrefix
+        value: /
+    timeouts:
+      backendRequest: 24h
+      request: 24h
```
@danehans why is this required?

LiorLieberman commented on site-src/guides/index.md:
````diff
@@ -56,55 +55,126 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+         ```
+
+=== "Istio"
+
````
can we add a note that this is experimental and users should NOT use it for any production environment. Also that the implementation/UX is likely to change as we iterate on the project.

Thoughts? 
@robscott @howardjohn

nicolexin replied:
Updated, PTAL and let me know if we need any updates!

LiorLieberman reviewed (commented):
Thanks for putting this together @nicolexin !

robscott commented on config/manifests/gateway/gke/gateway.yaml:
```diff
@@ -0,0 +1,15 @@
+kind: Gateway
+apiVersion: gateway.networking.k8s.io/v1beta1
```
```suggestion
apiVersion: gateway.networking.k8s.io/v1
```

robscott commented on config/manifests/gateway/httproute.yaml:
```diff
@@ -0,0 +1,19 @@
+    - group: inference.networking.x-k8s.io
+      kind: InferencePool
+      name: vllm-llama2-7b
+      weight: 1
```
This is the default value, so we can leave it out.
```suggestion
```

robscott commented on config/manifests/gateway/gke/gateway.yaml:
```diff
@@ -0,0 +1,15 @@
+ - name: http
+   port: 80
+   protocol: HTTP
+   allowedRoutes:
+     kinds:
+     - kind: HTTPRoute
+     namespaces:
+       from: All
```
I'd remove this from all Gateways because it's not actually necessary (assuming examples are all in same namespace). Nothing wrong with this, just hoping we can keep examples as concise as possible.

nicolexin replied:
Make sense, removed from all Gateways :)

robscott commented on site-src/guides/index.md:
```diff
@@ -56,55 +55,136 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+
+      1. Install Istio
+      
+         Please follow the [Istio installation guide](https://istio.io/latest/docs/setup/install/).
```
@LiorLieberman don't users need to use a specific branch of Istio to get this functionality?

LiorLieberman replied:
They do, I left that comment above, waiting for a release.

LiorLieberman replied:
after release is there, I would just point to an installation guide with overriding the images tag.

robscott reviewed (commented):
Thanks @nicolexin, this is a great update!

christian-posta commented on site-src/guides/index.md:
````diff
@@ -53,58 +52,146 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+      1. Install Kgateway CRDs
+
+         ```bash
+         helm upgrade -i --create-namespace --namespace kgateway-system --version v2.0.0-main kgateway-crds https://github.com/danehans/toolbox/raw/refs/heads/main/charts/338661f3be-kgateway-crds-1.0.1-dev.tgz
````
So this link won't work when there are updates to this repo. We should pin this to a specific commit. I'm not sure which is the right one just yet, let me dig into this and get back to you tomorrow.

danehans replied:
Yes, we'll want to use the v0.2.0 tag when the release is cut and do the same with the Helm repository.

ahg-g replied:
should we proceed with this PR and create an issue to followup with the fix?

robscott replied:
@christian-posta do you have a rough time on Friday in mind for Kgateway release?

christian-posta replied:
Yeah, RC-1 will come out later tonight, and Final should be cut Friday/Saturday. I can speed it up and get it Friday afternoon if things look good. 

I think my plan will be to get the inference extension verified right now on kgateway 2.0-main and then just update the instructions here as though there's a v2.0.0 tag. If we have any issues after that's cut and this goes out, we'll just submit a new PR to fix it.

christian-posta replied:
Okay so i have the following instructions working on my end, and we can update this guide to reflect this:

> kubectl apply -f llm/gpu-deployment.yaml

```bash
VERSION=v0.2.0
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/releases/download/$VERSION/manifests.yaml
```
> kubectl apply -f inference/inferencemodel.yaml
> kubectl apply -f inference/inferencepool.yaml

> helm upgrade -i --create-namespace --namespace kgateway-system --version v2.0.0 kgateway-crds oci://cr.kgateway.dev/kgateway-dev/charts/kgateway-crds

> helm upgrade -i --namespace kgateway-system --version v2.0.0 kgateway oci://cr.kgateway.dev/kgateway-dev/charts/kgateway \
  --set inferenceExtension.enabled=true 

> kubectl apply -f inference/gateway.yaml

> kubectl apply -f inference/httproute.yaml

try and call:
./call-gateway.sh

nicolexin replied:
I have now updated the Kgateway instructions, PTAL!

ahg-g commented on config/manifests/gateway/gke/healthcheck.yaml:
```diff
@@ -0,0 +1,16 @@
+kind: HealthCheckPolicy
```
Can we add this to the EPP helm chart as a gated by a provider flag? can be a followup, but wondering if you plan to do it?

nicolexin replied:
Make sense! I am not familiar with the EPP helm chart, it might make more sense for someone else to work on it as a follow up.

ahg-g commented on config/manifests/gateway/traffic_policy.yaml:
```diff
@@ -1,16 +0,0 @@
-apiVersion: gateway.envoyproxy.io/v1alpha1
```
What is the alternative to this in kgateway, istio and gke? @kfswain fyi

nicolexin replied:
I think for GKE we have GCPBackendPolicy where we can configure the timeout and maxRatePerEndpoint, not sure about kgateway and Istio.
cc @LiorLieberman @danehans

ahg-g replied:
>I think for GKE we have GCPBackendPolicy where we can configure the timeout and maxRatePerEndpoint

Do we want to add that? @robscott wdyt?

nicolexin replied:
Btw - The default timeout is only 30 sec according to https://cloud.google.com/kubernetes-engine/docs/how-to/configure-gateway-resources#configure_timeout. I think it make sense to at least add the timeout, @robscott wdyt? What would be a good default value to start with?

LiorLieberman replied:
IIUC the timeout is going to be the same as [HTTPRoute.timeout](https://gateway-api.sigs.k8s.io/reference/spec/#httproutetimeouts)? if yes, we can add it to the route (though we need to make sure all implementations support it)

ahg-g replied:
can we set it to 90 seconds? @smarterclayton any recommendations on the timeout?

smarterclayton replied:
You need to set timeout to `longest_request_at_peak_load + maximum_queue_wait_expected`.  To be really accurate it's going to be something like `max_context_length_of_model / number_of_batches_per_second_for_your_model_when_fully_compute_bound`, which for llama3 on h100 is 128k / 30 = 4200 seconds.

I have been setting it to 300s in my testing because I am running medium length responses (~average 200, max 10k).  It's going to be hard to pick a good out of the box and we will almost certainly want to come back and change it.

ahg-g commented on site-src/guides/index.md:
````diff
@@ -53,58 +52,146 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
    ```bash
-   kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/crd/bases/inference.networking.x-k8s.io_inferencepools.yaml
-   kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/crd/bases/inference.networking.x-k8s.io_inferencemodels.yaml
+   kubectl apply -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/releases/download/v0.2.0/manifests.yaml
````
Recommend to define a VERSION variable so that we can use it here and potentially in other places (the helm chart version), this makes it easy to update on each release.

Also, we need to keep this to show how to install the dev version (perhaps as a second tab), the instruction would be (instead of the two kubectl apply commands we have now):
```
kubectl apply -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/config/crd
```

nicolexin replied:
updated with 2 tabs, Latest release vs Dev version!

ahg-g commented on site-src/guides/index.md:
```diff
@@ -128,12 +215,13 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
-   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/traffic_policy.yaml --ignore-not-found
-   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/extension_policy.yaml --ignore-not-found
-   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/patch_policy.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/gateway.yaml --ignore-not-found
```
Can we please split it into three please: 
- `Uninstall the InferencePool` (uninstalls the InferencePool and InferenceModels)
- `Uninstall the gateway` (uninstalls the httproute and gateway)
- `Uninstall the APIs` (uninstalls the CRDs)

ahg-g reviewed (commented):
This is awesome!

LiorLieberman commented on config/manifests/inferencepool.yaml:
@nicolexin @ahg-g should we split this to different files? or change the file name? its not only the inferencepool, it is the EPP infra as well

ahg-g replied:
we are thinking of the epp as the implementation of the pool. I recommend to followup and remove all those manifests and just use helm.

LiorLieberman replied:
the name is still confusing. as its not only an inferencepool. I left some thoughts regarding helm here - https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/577/files#r2017600682

LiorLieberman replied:
most of the debate is on the other thread. For this yaml - I think just the name is confusing. I would call it something else other than inferencepool cause to me inferencepool means inferencepool. Maybe `inferencepool-deployment.yaml` ?

This is very nit though and shouldn't block ofc

nicolexin replied:
How about inferencepool-resources.yaml? I feel like inferencepool-deployment.yaml could be misleading too as it  may just refer to the underlying model server deployments.

nicolexin replied:
This is now renamed to inferencepool-resources.yaml.

LiorLieberman commented on config/manifests/gateway/istio/destination-rule.yaml:
```diff
@@ -0,0 +1,10 @@
+metadata:
+  name: epp-insecure-tls
+spec:
+  host: vllm-llama2-7b-epp.default.svc.cluster.local
```
note that this is currently configured to the `default` namespace. Do we have a place in the guide where we are instruct to configure the namespace? If no, we can add remove the suffix here and just do `host: vllm-llama2-7b-epp` and this would assume this yaml is applied to the same NS as the EPP service.

cc: @robscott @nicolexin @ahg-g

ahg-g replied:
I would like to move all of those manifests to be managed by helm

LiorLieberman replied:
I dont think you need a helm chart just for 3 manifests. We could iterate on this in the future, but I dont think its a good idea to have helm chart that covers all providers, and is maintained here. Doesn't feel right.

LiorLieberman replied:
because you dont want to manage provider-specific resources in a chart. Imagine what would have happened if we were to do that with Gateway api - and have every provider put their resources in there. There are some extensions, third-party apis that are added and will be added to support functionality that are not yet standardized.

Check out the timeout example - not we would need to do a if statement to see if an implementation supports timeout add it, if not, remove it. 

Similarly - you want to let implementation to iterate on their charts on their own - and not releasing new versions from the main repo here.

LiorLieberman replied:
co-locating whats needed for **Standardized** inferencepool infra in a chart makes sense, (although its only 3 resources, so I'd argue having this in the same yaml file is also fine). But having provider specific things in that chart is the problem and can easily become un maintainable. I think every implementation should have their own chart, and we could publish some guidance here.

LiorLieberman replied:
We chatted offline, 

I really do not see value in getting provider specific things into helm. It does not really simplify the user experience imo. And its having this maintained here is going to be obsolete very soon and will make the chart not maintainable. Even if we add it only initially with warning with we are getting to proper deprecation pain in the near future. 

Anyway, its not in helm right now so we can talk more later.

LiorLieberman replied:
If we dont specify namespace in the guide - I think we should follow what I suggested and use short service name as I suggested in the first comment in this thread

nicolexin replied:
Namespace on the host is now removed.

LiorLieberman commented on site-src/guides/index.md:
````diff
@@ -52,59 +51,152 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+         ./istioctl install --set tag=$TAG --set hub=gcr.io/istio-testing
+         ```
+
+      1. If you run the Endpoint Picker (EPP) with TLS (with `--secureServing=true`), it is currently using a self-signed certificate 
+      and the gateway cannot successfully validate the CA signature and the SAN. Apply the destination rule to bypass verification as 
+      a temporary workaround. A better TLS implementation is being discussed in [Issue 582](https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/582).
````
I want this to indicate that this problem is not unique to istio. And other implementations are doing the same thing implicitly. Any ideas how it is best to communicate that? 

@robscott @danehans

robscott replied:
Maybe something like this?

```suggestion
      1. If you run the Endpoint Picker (EPP) with the `--secureServing` flag set to `true`, it is currently using a self-signed certificate. As a security measure, Istio does not trust self-signed certificates by default. As a temporary workaround, you can apply the destination rule to bypass TLS verification for EPP. A more secure TLS implementation in EPP is being discussed in [Issue 582](https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/582).
```

As far as I can tell, this setting is not enabled by default, so it may be worth hiding this in a [collapsible admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks).

LiorLieberman replied:
It is the default - https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/61125a818d5a26addb3acf7a9b099cba82976761/cmd/epp/main.go#L88-L89

LiorLieberman replied:
But, my comment is more about the fact that all implementations (keep me honest for gke) are configuring envoy to accept untrusted certificate (which is what DestRule is doing in this case) so only mentioning it here is slightly misleading

nicolexin replied:
Yes, TLS is enabled by default for EPP. I agree it is slightly misleading here as we only mention it for Istio. However, I also feel like it is redundant to say in other implementations that they do accept untrusted certificates so there is no AI required from the users, especially if a more secure TLS implementation is coming soon.

robscott replied:
@LiorLieberman My goal with the suggestion was to highlight that Istio's TLS verification is a positive/helpful feature. IMO, the only time neutral OSS docs should point out a shortcoming of an implementation is if that implementation is failing to do something required by the API, that's not the case here.

LiorLieberman replied:
I am good with this suggestion Rob, thanks.

I think its misleading to not include it only here as others dont see this when they are not on the istio tab. 

But happy to move forward with this as well if thats the preference.

LiorLieberman commented on site-src/guides/index.md:
```diff
@@ -52,59 +51,152 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
-   For high-traffic benchmarking you can apply this manifest to avoid any defaults that can cause timeouts/errors.
+=== "Istio"
+
+      Please note that this feature is currently in an experimental phase and is not intended for production use. 
+      The implementation and user experience are subject to changes as we continue to iterate on this project.
```
I believe this is not unique to istio, but we only include this in istio. Any recommendations how to rephrase re-anchor this comment? 

@kfswain @ahg-g @robscott @nicolexin ?

robscott replied:
I think it's still worth leaving a small Istio-specific note here, but I added a recommendation for an experimental disclaimer at the top of the guide. @LiorLieberman up to you on what this should say if we have an overall disclaimer above.

LiorLieberman replied:
Thanks, we can leave it like this then

christian-posta commented:
I've added my review of the instructions for kgateway and updated the thread above with the steps/commands. 

A few thoughts from me:

1. the gpu-deployment uses "vllm/vllm-openai:latest"; is that intended? I think we should pin to a specific version. 

https://github.com/nicolexin/gateway-api-inference-extension/blob/userguide/config/manifests/vllm/gpu-deployment.yaml#L17

2. this ClusterRoleBinding gives me an error when I apply it:

https://github.com/nicolexin/gateway-api-inference-extension/blob/userguide/config/manifests/inferencepool.yaml#L117

```
inferencepool.inference.networking.x-k8s.io/vllm-llama3-8b-instruct created
service/vllm-llama3-8b-instruct-epp created
deployment.apps/vllm-llama3-8b-instruct-epp created
clusterrole.rbac.authorization.k8s.io/pod-read unchanged
error: error validating "https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/inferencepool.yaml": error validating data: ValidationError(ClusterRoleBinding.roleRef): missing required field "apiGroup" in io.k8s.api.rbac.v1.RoleRef; if you choose to ignore these errors, turn validation off with --validate=false
```

3. With the startupProbe and agressive liveness/readiness I could not get the llm container to come up FYI. I understand why these settings are important -- we may need to fiddle with them a little to arrive at good defaults. Have others had any issues with this? I've had to back those out on my side to get the llms to work... 

Thanks!

robscott commented on config/manifests/gateway/kgateway/gateway.yaml:
```diff
@@ -0,0 +1,12 @@
+# Requires Kgateway 2.0.0 or greater.
+---
```
Nit: I think you can remove this from this manifest now and save it for the markdown docs.

nicolexin replied:
done :)

robscott commented on site-src/guides/index.md:
```diff
@@ -3,7 +3,6 @@
 This quickstart guide is intended for engineers familiar with k8s and model servers (vLLM in this instance). The goal of this guide is to get a first, single InferencePool up and running! 
```
```suggestion
This quickstart guide is intended for engineers familiar with k8s and model servers (vLLM in this instance). The goal of this guide is to get an Inference Gateway up and running! 
```

robscott replied:
Recommend adding another section above this:

```
??? example "Experimental"

    This project is still in an alpha state and breaking changes may occur in the future.
```

This relies on [mkdocs-material Admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/), something we haven't used in this project quite yet. It will need some corresponding [updates to mkdocs.yml](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#configuration).

nicolexin replied:
done :)

robscott commented on site-src/guides/index.md:
```diff
@@ -52,59 +51,152 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+      1. Requirements
+
+         - [Helm](https://helm.sh/docs/intro/install/) installed.
+         - Gateway API [CRDs](https://gateway-api.sigs.k8s.io/guides/#installing-gateway-api) installed.
```
@LiorLieberman I believe this should be part of the Istio guide as well?

LiorLieberman replied:
assuming this comment is anchored only to gateway-apis crds installation and not helm - yes. we do need that. 

and you do need that before the installing with istioctl

can also do this - but not great we are tying to version (1.2.1)
```
kubectl get crd gateways.gateway.networking.k8s.io &> /dev/null || \
  { kubectl kustomize "github.com/kubernetes-sigs/gateway-api/config/crd?ref=v1.2.1" | kubectl apply -f -; }
```

nicolexin replied:
I updated the instructions to include the installation of Gateway CRDs, same as Kgateway.

robscott commented on site-src/guides/index.md:
````diff
@@ -126,18 +218,33 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+   1. Uninstall the Gateway
+
+   ```bash
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/gateway.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/healthcheck.yaml --ignore-not-found
````
```suggestion
   kubectl delete -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke --ignore-not-found
```

nicolexin replied:
I don't have kustomize configurations for these gateway manifests, so this replacement won't work. We only have kustomize configurations for the CRDs here: https://github.com/nicolexin/gateway-api-inference-extension/tree/userguide/config/crd.

robscott commented on site-src/guides/index.md:
````diff
@@ -126,18 +218,33 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+   ```bash
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/gateway.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/healthcheck.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/istio/gateway.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/istio/destination-rule.yaml --ignore-not-found
````
```suggestion
   kubectl delete -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/istio --ignore-not-found
```

robscott commented on site-src/guides/index.md:
```diff
@@ -126,18 +218,33 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/gke/healthcheck.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/istio/gateway.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/istio/destination-rule.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/kgateway/gateway.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/httproute.yaml --ignore-not-found
```
```suggestion
   kubectl delete -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/gateway/kgateway --ignore-not-found
```

nicolexin replied:
Same as above, this replacement won't work here.

robscott commented on site-src/guides/index.md:
````diff
@@ -126,18 +218,33 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+   1. Uninstall the CRDs
+
+   ```bash
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/crd/bases/inference.networking.x-k8s.io_inferencepools.yaml --ignore-not-found
+   kubectl delete -f https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/crd/bases/inference.networking.x-k8s.io_inferencemodels.yaml --ignore-not-found
+   kubectl delete -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/config/crd --ignore-not-found
````
Uninstalling CRDs will also delete all CRs based on them, so can simplify to this:
```suggestion
   kubectl delete -k https://github.com/kubernetes-sigs/gateway-api-inference-extension/config/crd --ignore-not-found
```

nicolexin replied:
done :)

robscott reviewed (commented):
Thanks @nicolexin, this looks great!

robscott reviewed (commented):
Thanks @nicolexin, this looks great!

robscott commented on site-src/guides/index.md:
````diff
@@ -52,59 +54,149 @@ This quickstart guide is intended for engineers familiar with k8s and model serv
+         ./istioctl install --set tag=$TAG --set hub=gcr.io/istio-testing
+         ```
+
+      1. If you run the Endpoint Picker (EPP) with the `--secureServing` flag set to `true`, it is currently using a self-signed certificate. As a security measure, Istio does not trust self-signed certificates by default. As a temporary workaround, you can apply the destination rule to bypass TLS verification for EPP. A more secure TLS implementation in EPP is being discussed in [Issue 582](http ...
````
Maybe add this for clarity:
```suggestion
      1. If you run the Endpoint Picker (EPP) with the `--secureServing` flag set to `true` (the default mode), it is currently using a self-signed certificate. As a security measure, Istio does not trust self-signed certificates by default. As a temporary workaround, you can apply the destination rule to bypass TLS verification for EPP. A more secure TLS implementation in EPP is being discussed in [Issue 582](https://github.com/kubernetes-sigs/gateway-api-inference-extension/issues/582).
```

robscott reviewed (commented):
Thanks @nicolexin!

LiorLieberman reviewed (approved):
Thanks for all the work here @nicolexin! overall LGTM. 
minor nits on some threads (nothing is actionable for now probably)

/lgtm

robscott reviewed (approved):
Thanks @nicolexin!

/lgtm

nicolexin commented:
/assign kfswain

ahg-g commented:
/approve

nicolexin commented:
> I've added my review of the instructions for kgateway and updated the thread above with the steps/commands.
> 
> A few thoughts from me:
> 
> 1. the gpu-deployment uses "vllm/vllm-openai:latest"; is that intended? I think we should pin to a specific version.
> 
> https://github.com/nicolexin/gateway-api-inference-extension/blob/userguide/config/manifests/vllm/gpu-deployment.yaml#L17
> 
> 2. this ClusterRoleBinding gives me an error when I apply it:
> 
> https://github.com/nicolexin/gateway-api-inference-extension/blob/userguide/config/manifests/inferencepool.yaml#L117
> 
> ```
> inferencepool.inference.networking.x-k8s.io/vllm-llama3-8b-instruct created
> service/vllm-llama3-8b-instruct-epp created
> deployment.apps/vllm-llama3-8b-instruct-epp created
> clusterrole.rbac.authorization.k8s.io/pod-read unchanged
> error: error validating "https://github.com/kubernetes-sigs/gateway-api-inference-extension/raw/main/config/manifests/inferencepool.yaml": error validating data: ValidationError(ClusterRoleBinding.roleRef): missing required field "apiGroup" in io.k8s.api.rbac.v1.RoleRef; if you choose to ignore these errors, turn validation off with --validate=false
> ```
> 
> 3. With the startupProbe and agressive liveness/readiness I could not get the llm container to come up FYI. I understand why these settings are important -- we may need to fiddle with them a little to arrive at good defaults. Have others had any issues with this? I've had to back those out on my side to get the llms to work...
> 
> Thanks!

I did run through the guide with a GKE cluster end to end and I have no issues applying the ClusterRoleBinding and the vLLM deployments. 
@LiorLieberman - did you run into any of the issues above?

kfswain commented:
@christian-posta 
> 1. the gpu-deployment uses "vllm/vllm-openai:latest"; is that intended? I think we should pin to a specific version.

Yeah, we mark a specific branch in our version branches. Granted that doesn't make it to our site (we only host main). We may need to break out version specific guides. 

> 2. this ClusterRoleBinding gives me an error when I apply it:

Interesting, I think that's just a validation error, and they RoleBinding should still exist afaik? I did omit `apiGroup` but I think it defaults to rbac.authorization.k8s.io/v1

> 3. With the startupProbe and agressive liveness/readiness I could not get the llm container to come up FYI. I understand why these settings are important -- we may need to fiddle with them a little to arrive at good defaults. Have others had any issues with this? I've had to back those out on my side to get the llms to work...

They work for me. I'm using A100s its possible we need to have a disclaimer that they are tuned for A100 machines. LMK

kfswain commented:
Thanks @nicolexin!!! RIP xDS Surgery, you won't be missed :P
